src/FFmpegWriter.cpp

```diff
@@ -1525,11 +1525,23 @@ void FFmpegWriter::write_audio_packets(bool is_final) {
 			// Calculate total samples
 			total_frame_samples = samples_in_frame * channels_in_frame;
 
-			// Translate audio sample values back to 16 bit integers
-			for (int s = 0; s < total_frame_samples; s++, frame_position++)
-				// Translate sample value and copy into buffer
-				all_queued_samples[frame_position] = int(frame_samples_float[s] * (1 << 15));
-
+			// Translate audio sample values back to 16 bit integers with saturation
+			float valF;
+			int16_t conv;
+			const int16_t max16 = 32767;
+			const int16_t min16 = -32768;
+			for (int s = 0; s < total_frame_samples; s++, frame_position++) {
+				valF = frame_samples_float[s] * (1 << 15);
+				if (valF > max16)
+					conv = max16;
+				else if (valF < min16)
+					conv = min16;
+				else
+					conv = int(valF + 32768.5) - 32768; // +0.5 is for rounding
+
+				// Copy into buffer
+				all_queued_samples[frame_position] = conv;
+			}
```

Comment:
Let's try this again, with code I've actually tested first:

This can be implemented without the temporary variables, using `std::min()`, `std::max()`, and the `INT16_MIN` / `INT16_MAX` typedefs. e.g:
```suggestion
			for (int s = 0; s < total_frame_samples; s++, frame_position++) {
				all_queued_samples[frame_position] = std::min(INT16_MAX,
				    std::max(INT16_MIN, int(frame_samples_float[s] * (1 << 15)))
				);
			}
```
(Using `(1 << 15)` instead of `INT16_MAX` does make a difference when `frame_samples_float[s]` is exactly `-1.0`, turns out.)

Reply:
Or, with the rounding (which does seem like a good idea):
```c++
all_queued_samples[frame_position] = std::min(INT16_MAX,
    std::max(INT16_MIN, (int(frame_samples_float[s] * (1 << 15) + 32768.5) - 32768))
);

Reply:
My code may has better optimization by compiler for branching prediction routines.

Reply:
Eh. I'm _surprised_ that it doesn't optimize better, but I'll buy that I guess. (You tested even with `-O2` / `-DCMAKE_BUILD_TYPE=Release`, and it made a measurable difference?)

Normally the temporary variables would make me nervous about multithreading, but like I said this entire function is already marked non-threadsafe so that's a moot point. OK, works for me.